### DISCUSSION TITLE
[New] Mesurer les espaces de stockage

### DISCUSSION THREAD

**Discussion by ddebarbieux** [cnumr/best-practices#116] [Bonnes pratiques] [answered]
Avec l'expansion du Cloud, il est devenu facile et naturel de backuper régulièrement toutes les données d’un projet, y compris les bases de données de production. Ces backups sont ensuite conservés pendant des mois ou des années.

La mesure des espaces de stockage permet de rendre tangible ces phénomènes et ainsi de supprimer les stockages superflus.

**Comment by ACTLEM** [comment]
Le terme "Mesurer" fait plus référence à de l'outillage, toutefois cela rejoint l'idée de stratégie de cycle de vie de la donnée. J'avoue avoir essayé de trouver une recommandation (sachant que le format est déjà présent dans la BP 17), mais je n'ai trouvé que "Définir une durée de vie maximale suffisante de stockage"...qui ne me plaît pas trop ^^

**Reply by ddebarbieux** [reply]
Ma recommandation est justement d'utiliser les outils afin d'avoir une information concrète et prendre une décission. Ecrire une recommandation sur la prise de décission sans connaître le contexte me semble ambitieux.

Une solution peut être de parler de la "gestion de données" comme c'est fait dans cet article : [Je code: les bonnes pratiques en éco-conception de service numérique à destination des developpeurs](https://hal.archives-ouvertes.fr/hal-03009741/document) ?

**Reply by ACTLEM** [reply]
Comme BP que je propose:
- "Limiter au strict nécessaire le nombre de redondance des données" -> Après cela rejoint un peu la partie SLA (BP88 Adapter la qualité de service et le niveau de disponibilité). La règle de validation se reposera ici sur la qualité de service nécessaire et suffisante, comme le nombre de backup qui vont au délà des préconisations du SLA)...donc j'ai peur que cela fasse un peu doublon avec la BP88.
- "Réduire le volume de données stockées au nécessaire" -> Ici la notion de mesure pourrait apparaître dans la description et rendrait plus concrète la règle de validation. Règle de validation: le % de volume de données non nécessaire au fonctionnement du service <1%  (par exemple). A noter que les données légales doivent bien sûrs entrer dans les données nécessaires jusqu'à leur durée maximum.

**Comment by ACTLEM** [comment, marked as answer]
Note équipe projet: Création de deux bonnes pratiques autour du stockage de données:

- "Limiter au strict nécessaire le nombre de redondance des données" (Issue: https://github.com/cnumr/best-practices/issues/174)
- "Réduire le volume de données stockées au nécessaire" (Issue: https://github.com/cnumr/best-practices/issues/175)

**Reply by ddebarbieux** [reply]
@ACTLEM 

Du coup, je suis un peu perdu : 
* "Limiter au strict nécessaire le nombre de redondance des données" est redondante avec la BP#88 (sur les SLA)
* "Réduire le volume de données stockées au nécessaire" est une BP trop large. Elle reprise dans le bonne pratique sur les logs (BP#103), celle sur le cloud (BP#93), elle peut intégrer des questions sur les doublons entre les tables, sur les données mortes ou innactives...

En résumé, je ne sais pas où commencer :)

**Reply by ACTLEM** [reply]
Effectivement, ça semble trop redondant avec d'autres pratiques. Surtout la première qui rejoint la BP88. J'avoue que je ne me souviens plus les arguments qui ont poussé quand même à demande l'ajout. 
Pour la deuxième, il y a les logs et le stockage de données nécessaires au fonctionnement du service. Le stockage dans le cloud est une première étape. Les logs sont traitées dans une BP (103). Pour les données nécessaires au service stockées, je n'arrive pas à retrouver une BP qui en parle (on parle de stockage local, cache mais pas de réduction des données stockées et de la mise en place d'housekeeping).

**Comment by devcrafting** [comment]
Quid de la denormalisation, notamment par rapport à la discussion https://github.com/cnumr/best-practices/discussions/78 ? Pour éviter des calculs/des requêtes complexes en lecture, ne vaut-il pas mieux occuper plus d'espace avec de une forme de redondance (quand on denormalise) ?

**Reply by ACTLEM** [reply]
Ici on parle de données persistentes et non de données volatiles, mais je n'ai peut-être pas compris ta remarque.

**Reply by devcrafting** [reply]
Oui exactement quand je parle de denormalisation, c'est bien de la donnée potentiellement persistante. La question importante à mon avis est de savoir quel est le ratio de lecture/écriture (en ordre de grandeur). Quand il est de 1000 ou 10000 pour 1 (assez classique, voire souvent beaucoup plus élevé), il faut pouvoir comparer l'impact d'un stockage de données denormalisées pour limiter l'impact des lectures VS des lectures basées sur des requêtes très consommatrices.

J'ai un exemple où je n'ai travaillé que sur la partie performance (il y a quelques années) sur un catalogue e-commerce. L'existant était basé sur des requêtes ORM avec un cache, en denormalisant (et donc en occupant plus d'espace disque) nous avons multiplié par 10 la capacité de tenue en charge (si on résonne à capacité constante, 10 fois moins de ressources serveur - CPU surtout + un effet sur la RAM).

**Reply by ACTLEM** [reply]
Tu proposerais quoi comme BP sur ce sujet (ou ajouterais un complément à une BP)?
